### ISSUE TITLE
Composed sync/async filters don't work

### ISSUE THREAD

**Issue by datokrat** [antoniogarrote/rdfstore-js#125]
The query in this file will fail using the current mix of synchronous and asynchronous filters:
https://github.com/disco-network/rdfstore-js/commit/577f4b9a8702846cb2e0225e956e1397800cb9e1

It contains the following filter consisting of an asynchronously evaluated `EXISTS` expression composed with a synchronously evaluated filter `?p = 23`.
`FILTER (?p = 23 || EXISTS { ?x ns:price 42 })`

It seems that the problem is that all the asynchronous work necessary for `EXISTS` clauses happens in `preprocessExistentialFilters`, a function only processing the filter clauses in the highest level instead of recursively processing the parts such a composed expression as above. Hence, our `EXISTS` clause won't be preprocessed.

Some weeks ago I managed solving issue #115 by making all filters async instead of introducing `preprocessExistentialFilters` to outsource all the async code - and it seems that this problem vanishes there. It shouldn't take a long time for me to update those lines of code to present them as possible fix - or do you prefer a solution with the filters working sync?

**Comment by antoniogarrote** [comment]
Hi @datokrat!

Thanks for finding the problem with the nested EXISTS filters.

I found the problem with these filters requiring async callbacks in f8c2127a81295d0a27d3218d591b3aa1e853f792 but I overlooked complex filter expressions.

My thinking about the issue is that EXISTS are the only filters requiring async handlers. I was thinking about trying to rewrite the filtering logic in an async fashion as you did, but it's a massive change and I would prefer saving a few calls in the stack.

I have added a fix in 618a37d1e2577e926d7d35d8596cfa9a5da8b3b0 that should extend my original solution to complex filters.

I have also added as much of your commit as possible, like the test cases (thanks!) to the project, but for the time being I really prefer this approach unless we hit a wall somewhere. Let's save your code thought just in case.

Thanks again for the time spent on this, take a look and see if you find any problems.
